### PR TITLE
OnDiskBitmap palette getter/setters

### PR DESCRIPTION
**Pull request by kmatch98**
This is an addition to add color palette getters/setters for the OnDiskBitmap.  This will allow for palete animations/color cycling for bitmaps without having to load them into RAM.

Also, this adds a flag that provides an option to force refresh of an OnDiskBitmap whenever the palette is modified.

-------------
This work was spurred out of an exploration of animated icons.  @FoamyGuy demonstrated that using OnDiskBitmap allowed loading of significantly more images for the "TouchDeck" project, however currently there is no way of adjusting the color palette.  This addition allows modifications of an OnDiskBitmap's color palette to achieve these types of color cycling animations.

This capability probably has a relative small set of use cases, so if feedback is that this should be put on hold, I'm perfectly fine to park this one, and it will reside here for posterity.

-------
This allows OnDiskBitmap to do these kinds of palette animations:
Adafruit_logo  - OnDiskBitmap palette cycling

### CODE REVIEW

File: locale/circuitpython.pot [hunk 1]
```diff
@@ -3950,6 +3966,10 @@ msgstr ""
 msgid "syntax error in uctypes descriptor"
 msgstr ""
 
+#: shared-module/displayio/OnDiskBitmap.c
+msgid "this OnDiskBitmap is raw color, not indexed"
```
Comment (tannewt) [inline locale/circuitpython.pot:3970]: Unify this with the similar message above.

**Review by tannewt — CHANGES_REQUESTED** [review]
Could we just have a `Palette` object available as `.palette` instead? That would allow for transparency and such.

In fact, maybe we should switch to returning the pixel index here so that it's truly a value. Then TileGrid could map that to a color with its pixel_shader (like odb.palette).

**Comment by kmatch98** [comment]
> Could we just have a `Palette` object available as `.palette` instead? That would allow for transparency and such.
> 
> In fact, maybe we should switch to returning the pixel index here so that it's truly a value. Then TileGrid could map that to a color with its pixel_shader (like odb.palette).

To be perfectly honest I couldn’t decipher the syntax to make something act as a `.palette` getter and setter property versus just a plain function. Can you point to a good example?  This may be moot if we redefine this to be a palette object. 

As for whether it should be redefined as a normal palette object, I’m not sure I understand how this OnDiskBitmap works with ColorConverter, so I’m unsure if making it into a palette object will cause other issues. But I speculate that this OnDiskBitmap was designed primarily for non-color-indexed bitmaps, so a palette is only warranted for the indexed flavor. Maybe it should only create a palette for indexed bitmaps, but then ColorConverter will have to be modified to take a Palette object. 

Your suggestions and clarifications are welcome.

**Comment by tannewt** [comment]
> To be perfectly honest I couldn’t decipher the syntax to make something act as a `.palette` getter and setter property versus just a plain function. Can you point to a good example? This may be moot if we redefine this to be a palette object.

Sure! The key thing is that there is an intermediate "property" object that stores pointers to the getter and setter functions. See: https://github.com/adafruit/circuitpython/blob/main/shared-bindings/digitalio/DigitalInOut.c#L224 (This is how properties work internally.)

> As for whether it should be redefined as a normal palette object, I’m not sure I understand how this OnDiskBitmap works with ColorConverter, so I’m unsure if making it into a palette object will cause other issues. But I speculate that this OnDiskBitmap was designed primarily for non-color-indexed bitmaps, so a palette is only warranted for the indexed flavor.

Yup, I think `palette` can be `None` in the non-indexed case.

> Maybe it should only create a palette for indexed bitmaps, but then ColorConverter will have to be modified to take a Palette object.

You don't need ColorConverter if you have a Palette. ColorConverter is just there to change 24 bit "values" (aka colors) into 16 bit colors. Palette does this conversion already internally for its 24 bit colors.

The main change I think is: https://github.com/adafruit/circuitpython/blob/main/shared-module/displayio/OnDiskBitmap.c#L159 It should return the index and then the external palette can map it.

**Comment by jposada202020** [comment]
Interesting

**Comment by kmatch98** [comment]
@jepler I think [your work will supercede or obsolete this PR](https://github.com/adafruit/circuitpython/issues/4636).  I'll close this one for now and await to see if your additions also provide this capability.

**Pull request closed without merging**